src/test/datascience/notebook/executionService.mock.vscode.test.ts

```diff
@@ -0,0 +1,221 @@
+            observableResult.asObservable() as any
+        );
+        const kernelSocket = new Subject<KernelSocketInformation>();
+        kernelSocket.next();
+        kernelSocket.complete();
+        when(inotebook.kernelSocket).thenReturn(kernelSocket.asObservable());
+        const statusChange = ServerStatus.Idle;
+        when(inotebook.status).thenReturn(statusChange);
+        const statusEvent = new EventEmitter<ServerStatus>();
+        disposables.push(statusEvent);
+        when(inotebook.onSessionStatusChanged).thenReturn(statusEvent.event);
+        when(inotebook.setLaunchingFile(anything())).thenResolve();
+        when(inotebook.requestKernelInfo()).thenResolve({
+            channel: 'shell',
+            content: {
+                banner: '',
+                help_links: [],
+                implementation: '',
+                implementation_version: '',
+                language_info: { name: '', version: '' },
+                protocol_version: '',
+                status: 'ok'
+            },
+            header: {
+                date: '',
+                msg_id: '',
+                msg_type: 'kernel_info_reply',
+                session: '',
+                username: '',
+                version: ''
+            },
+            metadata: {},
+            parent_header: {
+                date: '',
+                msg_id: '',
+                msg_type: 'kernel_info_request',
+                session: '',
+                username: '',
+                version: ''
+            }
+        });
+        when(inotebook.waitForIdle(anything())).thenResolve();
+        when(inotebook.session).thenReturn(session);
+        return kernelProvider.getOrCreate(notebook, {
+            controller: controller.controller,
+            resourceUri: notebook.uri,
+            metadata: controller.connection
+        });
+    }
+    test('Execute cell', async () => {
+        await insertCodeCell('print("123412341234")', { index: 0 });
+        const cell = vscodeNotebook.activeNotebookEditor?.document.cellAt(0)!;
+
+        const nbCell: nbformat.ICodeCell = {
+            cell_type: 'code',
+            execution_count: 1,
+            metadata: {},
+            outputs: [<nbformat.IStream>{ name: 'stdout', output_type: 'stream', text: '123412341234' }],
+            source: 'print("123412341234")'
+        };
+        const session = new MockJupyterSession([nbCell], 0);
+        const kernel = createKernelWithMockJupyterSession(cell.notebook, session);
+
+        // Wait till execution count changes and status is success.
+        await Promise.all([
+            kernel?.executeCell(cell),
+            waitForTextOutput(cell, '123412341234', 0, false),
+            waitForExecutionCompletedSuccessfully(cell)
+        ]);
+    });
+    test('Execute a cell that prints and clears the output 100s of times (once each iteration)', async () => {
```

Comment:
Should we add some sort of perf analysis to this one? Jupyter is out of the loop so all perf problems should be in the UI, right?

Reply:
I want to do that in a separate PR, 
* THis PR is large as is
* Got some more perf fixes (unnecessary serialization/deserialization)
* Perf related to string concatenation (handling streams & ANSI codes)
* Test, tests, tests. test